### PR TITLE
Add PSync consumer removal

### PR DESCRIPTION
**Pull request by jrandallclark**
**Description (required)**

Add removal of PSync consumer and fixes bug in SendFriendRequest activity

**Tests performed (required)**

Tested on Motox

### CODE REVIEW

File: psync/src/main/cpp/psync-wrapper.cpp [hunk 1]
```diff
@@ -375,10 +376,19 @@ JNIEXPORT jboolean JNICALL Java_net_named_1data_jni_psync_PSync_00024Consumer_ad
...
 }
 
+JNIEXPORT void JNICALL Java_net_named_1data_jni_psync_PSync_00024Consumer_stop
+  (JNIEnv *env, jobject, jobject handle)
+{
+    ConsumerWrapper* consumerWrapper = (ConsumerWrapper*) env->GetDirectBufferAddress(handle);
+    consumerWrapper->consumer->stop();
+    delete consumerWrapper;
+    ALOG("%s", "Stopping consumer");
+
```
Comment (agawande) [inline psync/src/main/cpp/psync-wrapper.cpp]: Extra line can be removed.

File: psync/src/main/cpp/psync-wrapper.cpp [hunk 2]
```diff
@@ -375,10 +376,19 @@ JNIEXPORT jboolean JNICALL Java_net_named_1data_jni_psync_PSync_00024Consumer_ad
...
   return consumerWrapper->consumer->addSubscription(ndn::Name(env->GetStringUTFChars(prefix, nullptr)));
 }
 
+JNIEXPORT void JNICALL Java_net_named_1data_jni_psync_PSync_00024Consumer_stop
+  (JNIEnv *env, jobject, jobject handle)
+{
+    ConsumerWrapper* consumerWrapper = (ConsumerWrapper*) env->GetDirectBufferAddress(handle);
+    consumerWrapper->consumer->stop();
+    delete consumerWrapper;
+    ALOG("%s", "Stopping consumer");
```
Comment (agawande) [inline psync/src/main/cpp/psync-wrapper.cpp]: Maybe Stopped consumer.

File: psync/src/main/java/net/named_data/jni/psync/PSync.java [hunk 3]
```diff
@@ -79,8 +79,8 @@ public void removeUserNode(String prefix) {
             removeUserNode(m_buffer, prefix);
         }
 
-        public long getSeqNo(String prefix) {
-            return getSeqNo(m_buffer, prefix);
+        public void getSeqNo(String prefix) {
```
Comment (agawande) [inline psync/src/main/java/net/named_data/jni/psync/PSync.java]: Why is this changed to void? Someone might have use for seqNo even if we don't.

File: app/src/main/java/memphis/myapplication/psync/ConsumerManager.java [hunk 4]
```diff
@@ -165,13 +166,17 @@ public void createConsumer(String prefix) {
...
     }
 
     public void removeConsumer(String friend) {
         Timber.d("Removing " + friend + " as consumer");
-        // Does nothing yet. Need to add.
+        PSync.Consumer removedConsumer = consumers.get(friend);
+        consumers.remove(friend);
+        removedConsumer.stop();
+        removedConsumer = null;
+
```
Comment (agawande) [inline app/src/main/java/memphis/myapplication/psync/ConsumerManager.java]: Extra line.

File: app/src/main/java/memphis/myapplication/psync/ConsumerManager.java [hunk 5]
```diff
@@ -24,6 +24,7 @@
 
 import java.io.IOException;
 import java.util.ArrayList;
```
Comment (agawande) [inline app/src/main/java/memphis/myapplication/psync/ConsumerManager.java:26]: If ArrayList is no longer needed then can be removed.

File: app/src/main/java/memphis/myapplication/ViewFriendsActivity.java [hunk 6]
```diff
@@ -77,9 +77,9 @@ public void onClick(DialogInterface dialog, int which) {
                 // Temporary for testing purposes
                 user.setTrust(false);
                 realm.commitTransaction();
-                realm.close();
-                Globals.consumerManager.removeConsumer(friend);
+                Globals.consumerManager.removeConsumer(user.getNamespace());
                 Globals.producerManager.updateFriendsList();
+                realm.close();
```
Comment (agawande) [inline app/src/main/java/memphis/myapplication/ViewFriendsActivity.java:82]: Did we already switch to realm? Wonder if it is worth looking into using realm for jNDN PIB instead of sqlite (no need to cross-compile the sqlitejdbc driver after that).

Comment (agawande) [reply app/src/main/java/memphis/myapplication/ViewFriendsActivity.java:82]: Could be a task for Ritik if he has some free time.

**Review by agawande — COMMENTED** [review]
Just minor comments.

File: app/src/main/java/memphis/myapplication/ViewFriendsActivity.java (hunk 6 above)
Comment (dev-ritik) [reply app/src/main/java/memphis/myapplication/ViewFriendsActivity.java:82]: `Realm is a mobile database` Can it be extended on other platforms for use in jNDN?

Comment (jrandallclark) [reply app/src/main/java/memphis/myapplication/ViewFriendsActivity.java:82]: We're already using an Android-specific implementation of Sqlite3, so it's not unreasonable to consider using Realm. We could see if they want it added to jNDN or we could just write a class in npChat that extends PibImpl.

https://github.com/named-data/jndn/tree/master/src/net/named_data/jndn/security/pib

File: psync/src/main/cpp/psync-wrapper.cpp (hunk 2 above)
Comment (jrandallclark) [reply psync/src/main/cpp/psync-wrapper.cpp]: This was just forgotten debug code. I'll remove it.

File: psync/src/main/java/net/named_data/jni/psync/PSync.java (hunk 3 above)
Comment (jrandallclark) [reply psync/src/main/java/net/named_data/jni/psync/PSync.java]: Oops. I initially copied the files from the JNI PSync github and it's void there. Must have been something I fixed here but never changed there. After we merge this, I'll create a pull request there.

File: app/src/main/java/memphis/myapplication/psync/ConsumerManager.java (hunk 5 above)
Comment (jrandallclark) [reply app/src/main/java/memphis/myapplication/psync/ConsumerManager.java:26]: Still using it for hello callback.

**Comment by jrandallclark** [comment]
@agawande @dev-ritik

File: psync/src/main/java/net/named_data/jni/psync/PSync.java [hunk 7]
```diff
@@ -133,6 +133,8 @@ public long getSeqNo(String prefix) {
             return getSeqNo(m_buffer, prefix);
         }
 
+        public void stop() {stop(m_buffer);}
```
Comment (agawande) [inline psync/src/main/java/net/named_data/jni/psync/PSync.java]: Style should be same as below.

File: app/src/main/java/memphis/myapplication/psync/ConsumerManager.java [hunk 8]
```diff
@@ -165,13 +166,15 @@ public void createConsumer(String prefix) {
...
     }
 
     public void removeConsumer(String friend) {
         Timber.d("Removing " + friend + " as consumer");
-        // Does nothing yet. Need to add.
+        PSync.Consumer removedConsumer = consumers.get(friend);
+        consumers.remove(friend);
+        removedConsumer.stop();
+        removedConsumer = null;
         }
```
Comment (agawande) [inline app/src/main/java/memphis/myapplication/psync/ConsumerManager.java]: Bracket indentation is wrong.

File: app/src/main/java/memphis/myapplication/ViewFriendsActivity.java (hunk 6 above)
Comment (agawande) [reply app/src/main/java/memphis/myapplication/ViewFriendsActivity.java:82]: Okay lets use sqlitejdbc for now - see if you can review my pull request now (instructions are there). Should fix all libsqlitejdbc problems.

File: app/src/main/java/memphis/myapplication/SendFriendRequest.java [hunk 9]
```diff
@@ -59,23 +59,28 @@ protected void onCreate(Bundle savedInstanceState) {
...
+            }
+            for (String f : myFriendsList) {
+                mutualFriend = f.substring(f.lastIndexOf("/")+1);
+                friendsMessage = friendsMessage + f + "\n";
+            }
+            message.setText(friendsMessage);
         }
-        message.setText(friendsMessage);
+
         realm.close();
```
Comment (agawande) [inline app/src/main/java/memphis/myapplication/SendFriendRequest.java:83]: realm.close is also closed below? Something seems weird - is it the indentation?

**Review by agawande — COMMENTED** [review]
Few more minor comments.

File: app/src/main/java/memphis/myapplication/SendFriendRequest.java (hunk 9 above)
Comment (jrandallclark) [reply app/src/main/java/memphis/myapplication/SendFriendRequest.java:83]: The second realm.close call should be in the onClick scope.
